### ISSUE TITLE
Inconsistent hasSameSizeAs handling of collection & array arguments

### ISSUE THREAD

**Issue by chagmed** [assertj/assertj-core#2483]
#### Summary
`(myCollection).hasSameSizeAs(myArray)` assertion is OK
`(myArray).hasSameSizeAs(myCollection)` assertion doesn't compile

Is it possible to make `hasSameSizeAs` commutative with respect to collections and arrays?

#### Example

```java
final int[][] intPairs = new int[][]{{1, 10}, {1, 1}, {0, 1}};
List<RangeWithCompactConstructorForValidation> ranges =
    Arrays.stream(intPairs)
          .map(it -> RangeWithCompactConstructorForValidation
                  .make(it[0], it[1])).toList();

assertThat(ranges).hasSameSizeAs(intPairs);  // OK
assertThat(intPairs).hasSameSizeAs(ranges);  // won't compile
```

**Comment by scordio** [comment]
The last line doesn't compile because `assertThat(intPairs)` returns an [`Int2DArrayAssert`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/Int2DArrayAssert.html) and `hasSameSizeAs` is not available for two-dimensional assertions.

With #2296 / #2461 it will be possible to write:

```java
assertThat(intPairs).hasNumberOfRows(ranges.size());
```

I'm not sure how the use case could be improved further. I don't have a good proposal at the moment.

**Comment by scordio** [comment]
An option could be `hasNumberOfRowsAsSizeOf(Iterable)`, like:

```java
assertThat(intPairs).hasNumberOfRowsAsSizeOf(ranges);
```

but honestly I don't like it that much.

**Comment by joel-costigliola** [comment]
I agree with @scordio, `assertThat(intPairs).hasNumberOfRows(ranges.size());` looks reasonable to me and `assertThat(intPairs).hasNumberOfRowsAsSizeOf(ranges);` feels to specific.
